### PR TITLE
Fix/740

### PR DESCRIPTION
**Pull request by wind57**

### CODE REVIEW

File: spring-cloud-kubernetes-client-config/src/main/java/org/springframework/cloud/kubernetes/client/config/reload/KubernetesClientEventBasedConfigMapChangeDetector.java [hunk 1]
```diff
@@ -70,21 +70,21 @@ public class KubernetesClientEventBasedConfigMapChangeDetector extends Configura
 	private final ResourceEventHandler<V1ConfigMap> handler = new ResourceEventHandler<>() {
 
 		@Override
```
Comment (wind57) [inline spring-cloud-kubernetes-client-config/src/main/java/org/springframework/cloud/kubernetes/client/config/reload/KubernetesClientEventBasedConfigMapChangeDetector.java:72]: just renames here - nothing else

File: spring-cloud-kubernetes-client-config/src/main/java/org/springframework/cloud/kubernetes/client/config/reload/KubernetesClientEventBasedSecretsChangeDetector.java [hunk 2]
```diff
@@ -70,21 +70,21 @@ public class KubernetesClientEventBasedSecretsChangeDetector extends Configurati
 	private final ResourceEventHandler<V1Secret> handler = new ResourceEventHandler<>() {
 
 		@Override
-		public void onAdd(V1Secret obj) {
```
Comment (wind57) [inline spring-cloud-kubernetes-client-config/src/main/java/org/springframework/cloud/kubernetes/client/config/reload/KubernetesClientEventBasedSecretsChangeDetector.java:73]: renames only

File: spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/BusEventBasedConfigMapWatcherChangeDetector.java [hunk 3]
```diff
@@ -47,8 +47,8 @@ final class BusEventBasedConfigMapWatcherChangeDetector extends ConfigMapWatcher
 	}
 
 	@Override
-	public Mono<Void> triggerRefresh(KubernetesObject configMap) {
-		return busRefreshTrigger.triggerRefresh(configMap);
+	public Mono<Void> triggerRefresh(KubernetesObject configMap, String appName) {
```
Comment (wind57) [inline spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/BusEventBasedConfigMapWatcherChangeDetector.java:50]: we can have multiple apps based in a configMap/secret now

File: spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/WatcherUtil.java [hunk 4]
```diff
@@ -16,18 +16,25 @@
 
 package org.springframework.cloud.kubernetes.configuration.watcher;
```
Comment (wind57) [inline spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/WatcherUtil.java:18]: the main changes are in this class.

File: spring-cloud-kubernetes-integration-tests/spring-cloud-kubernetes-client-configmap-event-reload-multiple-apps/pom.xml [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+	<parent>
+		<artifactId>spring-cloud-kubernetes-integration-tests</artifactId>
+		<groupId>org.springframework.cloud</groupId>
+		<version>3.0.0-SNAPSHOT</version>
+	</parent>
+	<modelVersion>4.0.0</modelVersion>
+	<artifactId>spring-cloud-kubernetes-client-configmap-event-reload-multiple-apps</artifactId>
+	<packaging>pom</packaging>
+
+	<modules>
```
Comment (wind57) [inline spring-cloud-kubernetes-integration-tests/spring-cloud-kubernetes-client-configmap-event-reload-multiple-apps/pom.xml:14]: an integration test where two apps a re deployed, one configmap that has the new annotation in place: that refresh both of the apps

File: spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/WatcherUtil.java [hunk 6]
```diff
@@ -40,37 +46,93 @@ final class WatcherUtil {
...
-				try {
-					triggerRefresh.apply(kubernetesObject).subscribe();
-				}
-				catch (Throwable t) {
-					LOG.warn(t, "Error when refreshing ConfigMap " + name);
-				}
-			}, refreshDelay, TimeUnit.MILLISECONDS);
+			Set<String> apps = apps(kubernetesObject, annotationName);
+
+			if (!apps.isEmpty()) {
```
Comment (wind57) [inline spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/WatcherUtil.java:60]: the crox of the change is here: if the annotation that says we should restart multiple apps is present - try to get apps from it, and send events based on those names

File: spring-cloud-kubernetes-integration-tests/spring-cloud-kubernetes-client-secrets-event-reload-multiple-apps/pom.xml [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+		<artifactId>spring-cloud-kubernetes-integration-tests</artifactId>
+		<groupId>org.springframework.cloud</groupId>
+		<version>3.0.0-SNAPSHOT</version>
+	</parent>
+
+	<modelVersion>4.0.0</modelVersion>
+	<artifactId>spring-cloud-kubernetes-client-secrets-event-reload-multiple-apps</artifactId>
+	<packaging>pom</packaging>
+
+	<modules>
```
Comment (wind57) [inline spring-cloud-kubernetes-integration-tests/spring-cloud-kubernetes-client-secrets-event-reload-multiple-apps/pom.xml:15]: two apps that react as a result of a secret being deployed with proper label and annotation. Just proves that we expect both applications to be updated

File: docs/src/main/asciidoc/spring-cloud-kubernetes-configuration-watcher.adoc [hunk 8]
```diff
@@ -112,11 +112,32 @@ Spring Cloud Kubernetes Configuration Watcher will react to changes in ConfigMap
 or any Secret with a label of `spring.cloud.kubernetes.secret` with the value `true`.  If the ConfigMap or Secret does not have either of those labels
 or the values of those labels is not `true` then any changes will be ignored.
 
-The labels Spring Cloud Kubernetes Configuration Watcher looks for on ConfigMaps and Secrets can be changed by setting
-`spring.cloud.kubernetes.configuration.watcher.configLabel` and `spring.cloud.kubernetes.configuration.watcher.secretLabel` respectively.
-
 If a change is made to a ConfigMap or Secret with valid labels then Spring Cloud Kubernetes Configuration Watcher will take the name of the ConfigMap or Secret
-and send a notification to the application with that name.
+and send a notification to the application with that name. This might not be enough for your use-case though, you could for example what to:
```
Comment (wind57) [inline docs/src/main/asciidoc/spring-cloud-kubernetes-configuration-watcher.adoc:116]: added documentation here for the new properties

**Comment by wind57** [comment]
@ryanjbaxter ready to be looked at

File: spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/WatcherUtil.java (hunk 6 above)
Comment (ryanjbaxter) [reply spring-cloud-kubernetes-controllers/spring-cloud-kubernetes-configuration-watcher/src/main/java/org/springframework/cloud/kubernetes/configuration/watcher/WatcherUtil.java:60]: We could simplify this a bit if we add the name to the set if its empty

**Review by ryanjbaxter — CHANGES_REQUESTED** [review]

**Comment by wind57** [comment]
@ryanjbaxter take a look please to see if I got your point correct. thank you

**Pull request merged**